### PR TITLE
PR comments from apache#34311

### PR DESCRIPTION
**Pull request by icexelloss**
Some comments from https://github.com/apache/arrow/pull/34311

### CODE REVIEW

**Comment by rtpsw** [comment]
@icexelloss, the proposed changes are mostly OK. There is a conflict to resolve. I also suspect some automated conflict-resolutions did not go well, because I see both `ResetAggregates` and `ResetKernelStates`. Do you want to resolve the conflict(s) or should I? As for renaming symbols, I'm fine with `RowSegmenter` but `SegmentPiece` is odd, as a segment already means a piece. How about `RowSegment` for consistency?

**Comment by icexelloss** [comment]
> I'm fine with RowSegmenter but SegmentPiece is odd, as a segment already means a piece. How about RowSegment for consistency?

I found the this confusing when first reading the code so I renamed this. Segment and segment piece are different concept. In my mind a segment is a set of rows with the same key and can span multiple batches. A segment piece a slice of a batch that belongs to a segment. I don't have strong opinions about the sames but if you want those are two things and we should be explicit about it. If what I describe as "segment piece" is what you call "segment" we need another term to represent "a collection of continuous segment that has the same key"

**Comment by icexelloss** [comment]
@rtpsw I named "Segment Piece" back to "Segment" and called the "a collection of continuous segment that has the same key" tentatively "segment group". Also resolved the conflict.

**Comment by rtpsw** [comment]
> A segment piece a slice of a batch that belongs to a segment.

I see. OK, then I guess the name is fine, just needs a docstring.

File: cpp/src/arrow/compute/row/grouper.h [hunk 1]
```diff
@@ -30,8 +30,15 @@
...
 namespace compute {
 
-/// \brief A segment of contiguous rows for grouping
-struct ARROW_EXPORT GroupingSegment {
+/// \brief A segment.
+/// A segment group is a chunk of continous rows that has the same segment key. (For
+/// example, in ordered time series processing, segment key can be "date", and a segment
+/// group can be all the rows that belong to the same date.) A segment group can span
+/// across multiple exec batches. A segment is a chunk of continous rows that has the same
+/// segment key within a given batch. When a ? span cross batches, it will have multiple
```
Comment (rtpsw) [inline cpp/src/arrow/compute/row/grouper.h]: Looks like the question mark here is a typo?

File: cpp/src/arrow/compute/row/grouper.h [hunk 2]
```diff
@@ -30,8 +30,15 @@
 namespace arrow {
 namespace compute {
 
-/// \brief A segment of contiguous rows for grouping
-struct ARROW_EXPORT GroupingSegment {
+/// \brief A segment.
+/// A segment group is a chunk of continous rows that has the same segment key. (For
```
Comment (rtpsw) [inline cpp/src/arrow/compute/row/grouper.h]: It's a bit confusing to start with a description of a segment group when the docstring is for segment. I'd say here something like "A segment is a chunk of contiguous rows that have the same segment key. Multiple contiguous segments having a common segment key form a segment group.".

Note that there is a typo: "that has" should be "that have".

File: cpp/src/arrow/compute/row/grouper.h [hunk 3]
```diff
@@ -30,8 +30,15 @@
...
+/// example, in ordered time series processing, segment key can be "date", and a segment
+/// group can be all the rows that belong to the same date.) A segment group can span
+/// across multiple exec batches. A segment is a chunk of continous rows that has the same
+/// segment key within a given batch. When a ? span cross batches, it will have multiple
+/// segments. A segment never span cross batches. The segment data structure only makes
```
Comment (rtpsw) [inline cpp/src/arrow/compute/row/grouper.h]: Typo: "span" should be "spans".

**Review by rtpsw — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/row/grouper.h (hunk 2 above)
Comment (icexelloss) [reply cpp/src/arrow/compute/row/grouper.h]: I think it doesn't really make sense to talk about segment without talking about segment group first - segment group is the top-level concept, segment is how we track batches within a segment group

File: cpp/src/arrow/compute/row/grouper.h (hunk 3 above)
Comment (icexelloss) [reply cpp/src/arrow/compute/row/grouper.h]: Good catch fixed.

File: cpp/src/arrow/compute/row/grouper.h (hunk 1 above)
Comment (icexelloss) [reply cpp/src/arrow/compute/row/grouper.h]: Fixed.

**Pull request merged**